### PR TITLE
[IoT Central] New API for 2022-07-31 release

### LINKED ISSUE #19694: [IoT Central] REST API Review (2022-07-31)
Service name
IoT Central

Key contact for the review
[hayang@microsoft.com](mailto:hayang@microsoft.com), [Avneet.Singh@microsoft.com](mailto:Avneet.Singh@microsoft.com)

Whether this is a new or existing API
Existing API

Whether or not your service already been deployed
No, TBD by end of July

A brief description of the material to be reviewed/changed:

Add Enrollment Group support
Add Scheduled Job Definition support
Add Jobs support
Add Device Group List Devices support
Add Device Relationships support
Add OData support from devices, device groups, and device templates.

[Swagger PR](https://github.com/Azure/azure-rest-api-specs/pull/19693)

### PR DESCRIPTION
**Pull request by yang-hai-feng**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

Changes from 2022-07-31:

- Add Enrollment Groups support from preview version.
- Add Jobs support from preview version
- Add Scheduled Jobs support from preview version
- Add Device Relationship support from preview version
- Add Device Group list devices support from preview version

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following appy to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Review by tjprescott — CHANGES_REQUESTED** [review]
This PR is not structured correctly to facilitate review. The first commit should copy over the old API Version. The 2nd commit should update the API Version in the new folder, and the third commit should reflect all new changes in this API Version.

**Comment by yang-hai-feng** [comment]
> This PR is not structured correctly to facilitate review. The first commit should copy over the old API Version. The 2nd commit should update the API Version in the new folder, and the third commit should reflect all new changes in this API Version.

Not sure why it had all the previous commits, but what I did is copy the 2022-05-31 for first commit:
image

From the first commits, the rest are adding new functions and lint rule fixes, let me know if this helps.

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json [hunk 1]
```diff
@@ -0,0 +1,6113 @@
...
+      "in": "path",
+      "name": "telemetryName",
+      "description": "Name of this device telemetry.",
+      "x-ms-parameter-location": "method",
+      "type": "string",
+      "required": true
+    },
+    "Top": {
+      "in": "query",
+      "name": "$top",
```
Comment (mikekistler) [inline specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json:6069]: Don't use "$" prefix on parameter names.

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json [hunk 2]
```diff
@@ -0,0 +1,6113 @@
...
+      "in": "path",
+      "name": "moduleName",
+      "description": "Name of the device module.",
+      "x-ms-parameter-location": "method",
+      "type": "string",
+      "required": true
+    },
+    "OrderBy": {
+      "in": "query",
+      "name": "$orderby",
```
Comment (mikekistler) [inline specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json:6017]: Don't use "$" prefix on parameter names.

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json [hunk 3]
```diff
@@ -0,0 +1,6113 @@
...
+            "name": "Secondary",
+            "description": "The entry of secondary certificate"
+          }
+        ]
+      },
+      "required": true
+    },
+    "Filter": {
+      "in": "query",
+      "name": "$filter",
```
Comment (mikekistler) [inline specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json:5991]: Don't use "$" prefix on parameter names.

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json [hunk 4]
```diff
@@ -0,0 +1,6113 @@
...
+        "summary": "Update a scheduled job via patch",
+        "description": "Update an existing scheduled job by ID.",
+        "x-ms-examples": {
+          "List jobs": {
+            "$ref": "./examples/scheduledjobs_update.json"
+          }
+        },
+        "consumes": [
+          "application/json",
+          "application/merge-patch+json"
```
Comment (mikekistler) [inline specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json:5565]: Patch should only accept "application/merge-patch+json"
```suggestion
          "application/merge-patch+json"
```

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json [hunk 5]
```diff
@@ -0,0 +1,6113 @@
...
+            "description": "Certificate for the signed verification code.",
+            "schema": {
+              "$ref": "#/definitions/X509VerificationCertificate"
+            },
+            "required": true
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Success"
```
Comment (mikekistler) [inline specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json:4674]: A 200 response should have a response body schema. If there is no body in the response then please use `204 NoContent`.

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json [hunk 6]
```diff
@@ -0,0 +1,6113 @@
...
+        "summary": "Update an enrollment group",
+        "description": "Update an enrollment group.",
+        "x-ms-examples": {
+          "Update an enrollment group by ID": {
+            "$ref": "./examples/enrollmentgroups_update.json"
+          }
+        },
+        "consumes": [
+          "application/json",
+          "application/merge-patch+json"
```
Comment (mikekistler) [inline specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json:4392]: Patch operation should only accept `application/merge-patch+json`.
```suggestion
          "application/merge-patch+json"
```

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json [hunk 7]
```diff
@@ -0,0 +1,6113 @@
...
+        "summary": "Patch a given relationship given the relationship ID and a given device ID.",
+        "description": "Update device relationship",
+        "x-ms-examples": {
+          "Update a device relationship": {
+            "$ref": "./examples/devices_relationships_update.json"
+          }
+        },
+        "consumes": [
+          "application/json",
+          "application/merge-patch+json"
```
Comment (mikekistler) [inline specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json:4131]: Patch should only accept `application/merge-patch+json`
```suggestion
          "application/merge-patch+json"
```

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json [hunk 8]
```diff
@@ -0,0 +1,6113 @@
...
+            }
+          }
+        }
+      }
+    },
+    "/devices/{deviceId}/relationships": {
+      "get": {
+        "operationId": "Devices_ListRelationships",
+        "summary": "Given the ID for an upstream device, will return the upstream and the downstream relationships associated with that gateway. These downstream relationships are only those associated with the direct downstream level (they don’t work recursively).",
+        "description": "List all relationships based on device ID",
```
Comment (mikekistler) [inline specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json:3992]: I think you want to mark this operation with `x-ms-pageable: true`

**Review by mikekistler — CHANGES_REQUESTED** [review]
I flagged a few style issues that I hope you can clean up.

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json (hunk 1 above)
Comment (yang-hai-feng) [reply specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json]: Removed $filter, $top, $orderby from July GA, will revise this and get preview on Sep.

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json (hunk 2 above)
Comment (yang-hai-feng) [reply specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json]: Removed $filter, $top, $orderby from July GA, will revise this and get preview on Sep.

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json (hunk 3 above)
Comment (yang-hai-feng) [reply specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json]: Removed $filter, $top, $orderby from July GA, will revise this and get preview on Sep.

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json (hunk 4 above)
Comment (yang-hai-feng) [reply specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json]: fixed

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json (hunk 6 above)
Comment (yang-hai-feng) [reply specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json]: Fixed

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json (hunk 8 above)
Comment (yang-hai-feng) [reply specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json:3962]: Added

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json (hunk 7 above)
Comment (yang-hai-feng) [reply specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json]: Fixed

**Comment by tjprescott** [comment]
@yang-hai-feng please refactor this PR to facilitate review. 

Commits should be structured like this:
1. Copies over the previous API Version to the new folder
2. Updates the APIVersion in the Swagger files in the new folder
3. Squashed changes that represent what actually has changed in this APIVerison.

**Comment by yang-hai-feng** [comment]
> @yang-hai-feng please refactor this PR to facilitate review.
> 
> Commits should be structured like this:
> 
> 1. Copies over the previous API Version to the new folder
> 2. Updates the APIVersion in the Swagger files in the new folder
> 3. Squashed changes that represent what actually has changed in this APIVerison.

Thanks, seems I missed Squashed changes part, let me try if I can update

**Comment by yang-hai-feng** [comment]
@tjprescott, I have squashed all commits to one after initial version from 2022-05-31 version, 
image

Please let me know if you still have issues.

**Comment by tjprescott** [comment]
@yang-hai-feng I still show 54 commits in this PR. Did you push your squashed changes?

**Comment by yang-hai-feng** [comment]
> @yang-hai-feng I still show 54 commits in this PR. Did you push your squashed changes?

@tjprescott Can't squash the previous released changes, I squashed all changes after I initialized the new version from 2022-05-31, you can select the latest commits: 
image

**Comment by tjprescott** [comment]
@yang-hai-feng I don't understand. There shouldn't be any "previously released changes" in this PR.

**Comment by yang-hai-feng** [comment]
> @yang-hai-feng I don't understand. There shouldn't be any "previously released changes" in this PR.

@tjprescott, if I don't re-fork the repo, keep using the same branch from fork and commit my changes, it will includes all commits that I submitted.

**Review by mikekistler — APPROVED** [review]
Looks good! 👍

**Review by weidongxu-microsoft — COMMENTED** [review]
Given the side and problem of the diff on the PR, I can only skim the swagger. Added a few comments.

One thing is that it seems payload of the PATCH is `object` without any properties.

**Review by markweitzel — APPROVED** [review]

**Review by tjprescott — APPROVED** [review]

File: specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json [hunk 9]
```diff
@@ -539,348 +765,2128 @@
...
+          {
+            "$ref": "#/parameters/DeviceGroupId"
+          },
+          {
+            "in": "body",
+            "name": "body",
+            "description": "Device group patch body.",
+            "schema": {
+              "type": "object"
+            },
```
[the remark this answers is not in the document]
Comment (yang-hai-feng) [reply specification/iotcentral/data-plane/Microsoft.IoTCentral/stable/2022-07-31/iotcentral.json:1967]: We used json merge patch, the patch body could be the Partial<Entity>, so there is no fixed type for that partial entity combination.

**Pull request merged**